rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+specific privileges on `db1`, and another role `myCustomRole` can be created on
+database `db2` with specific privileges on `db2`. They will be considered two
+different roles.
+
+Most built-in roles are available on all databases, while some built-in roles
+like `readAnyDatabase` can only be applied on the `admin` database.
+
+Therefore, when specifying database roles to assign for the user, it must be in
+the format of `<role-name>@<db-name>` to fully identify the role:
+
+```yaml
+kind: "role"
+version: "v6"
+metadata:
+  name: "example"
+spec:
+  options:
+    create_db_user_mode: keep
+  allow:
+    db_names:
+    - "db1"
+    - "db2"
+    - "db3"
+    db_roles: 
+    - "readAnyDatabase@admin"
+    - "readWrite@db2"
+    - "myCustomRole@db3"
+```
+
+Note that in the above example, even though the user is assigned role
+`readAnyDatabase@admin`, Teleport will block access to databases not in the
+`db_names`.
+
+### User accounts
+
+New users with name `CN=<teleport-username>` will be created on `$external`
+database to use X.509 authentication.
+
+All auto-provisioned users will be assigned role `teleport-auto-user@admin`.
+The `teleport-auto-user` is created on database `admin` with empty privileges.
+
+There is no built-in way to lock an user account in MongoDB, for deactivation
+purpose. As a workaround, the following `authenticationRestrictions` will be
+applied to the user account, in addition to stripping the roles:
+
+```json
+{
+  "updateUser": "CN=<teleport-username>",
+  "roles": [
+    { "role": "teleport-auto-user", "db": "admin" }
+  ],
+  "authenticationRestrictions": [
+    { "clientSource": ["0.0.0.0"] }
+  ]
+}
+```
+
+Limiting the `clientSource` effectively locks out the user from logging in.
+
+When re-activating the user, `clientSource` will be set to `["0.0.0.0/0"]`.
+
+### Finding active connections
+
+Command `currentOp` is used to find active connections for a specific user:
+```json
+{
+  "currentOp": true,
+  "$ownOps": false,
+  "$all": true,
+  "effectiveUsers": {
+    "$elemMatch": {
+      "user": "CN=<teleport-username>",
+      "db": "$external"
+    }
+  }
+}
+```
+
+## UX
+
+Database roles must be specified in format of `<role-name>@<db-name>`. See
+"Roles" section above for more details.
+
+## Security
+
+The admin user requires role `clusterMonitor@admin` and role
+`userAdminAnyDatabase@admin` to monitor connections and manage users.
+
+Auth restrictions `clientSource: ["0.0.0.0"]` is used to lock an user account
+when deactivated.
+
+See above sections for more details.
+
+## Performance
+
+As stated in [Issue
+#10950](https://github.com/gravitational/teleport/issues/10950), MongoDB
+clients usually spawn multiple connections to the server resulting multiple
+parallel database sessions on the Database Service. The number of connections
+can be limited using a smaller `maxPoolSize` (default 100) in the connection
+string, but Teleport does not have full control on this as it's specified from
```

Comment:
For "admin" connections, don't we have full control over this because Teleport basically _is_ the client?

Reply:
For `tsh db connect` yes. No for `tsh proxy db` or Teleport Connect.

We can change `tsh db connect` to use `maxPoolSize=1`. Even then, `mongosh` makes 3 connections.